File: backend/app/services/booking.py
from uuid import UUID
from typing import Any, Dict, Optional, List
from datetime import date
from app.models import BookingInDB, BookingCreate, BookingUpdate, BookingResponse
from app.models.enums import BookingStatus
from app.repository import BookingRepository
from app.utils.exceptions import (
    NotFoundException,
    ConflictException,
    InternalServerException,
)


def booking_in_db_to_response(booking: BookingInDB) -> BookingResponse:
    """Convert a BookingInDB object to a BookingResponse object"""
    return BookingResponse(**booking.model_dump())


class BookingService:
    def __init__(self, repo: BookingRepository) -> None:
        self.repo = repo

    async def create_booking(self, data: BookingCreate) -> BookingResponse:
        """"""
        if await self.repo.collides(data.technician_id, data.timeslot):
            raise ConflictException(
                f"Technician already has a booking at {str(data.timeslot)}, try a different time."
            )
        entry: Dict[str, Any] = {
            "client_id": data.client_id,
            "technician_id": data.technician_id,
            "service_name": data.service_name,
            "description": data.description,
            "booking_date": data.timeslot.slot_date,
            "start_time": data.timeslot.start_time,
            "end_time": data.timeslot.end_time,
            "location_name": data.location.location_name,
            "location": f"POINT({data.location.longitude} {data.location.latitude})"
        }
        booking: Optional[BookingInDB] = await self.repo.create(entry)
        if booking is None:
            raise InternalServerException("Error creating booking")
        return booking_in_db_to_response(booking)

    async def readone_booking(self, booking_id: UUID) -> BookingResponse:
        """"""
        booking: Optional[BookingInDB] = await self.repo.readone(booking_id)
        if booking is None:
            raise NotFoundException("Booking not found")
        return booking_in_db_to_response(booking)

    async def readall_bookings(
        self,
        client_id: Optional[UUID] = None,
        technician_id: Optional[UUID] = None,
        status: Optional[BookingStatus] = None,
        booking_date: Optional[date] = None,
        skip: int = 0,
        limit: int = 100,
    ) -> List[BookingResponse]:
        """"""
        bookings: List[BookingInDB] = await self.repo.readall(
            client_id,
            technician_id,
            status,
            booking_date,
            skip,
            limit,
        )
        return [booking_in_db_to_response(booking) for booking in bookings]

    async def update_booking(
        self, booking_id: UUID, data: BookingUpdate
    ) -> BookingResponse:
        """"""
        booking: Optional[BookingInDB] = await self.repo.readone(booking_id)
        if booking is None:
            raise NotFoundException("Booking not found")
        booking: Optional[BookingInDB] = await self.repo.update(
            booking_id, data.model_dump(exclude_unset=True)
        )
        if booking is None:
            raise InternalServerException("Error updating booking")
        return booking_in_db_to_response(booking)

    async def delete_booking(self, admin_id: UUID) -> bool:
        """"""
        return await self.repo.delete(admin_id)

File: frontend/src/lib/services/booking.ts
import axios from "axios";
import { BASE_API_URL } from "@/lib/constants";
import type { BookingStatus } from "@/lib/types/enums";
import type { BookingCreate, BookingUpdate, BookingResponse } from "@/lib/types/booking";

const URL = `${BASE_API_URL}/booking`;

export async function createBooking(data: BookingCreate): Promise<BookingResponse> {
    const response = await axios.post(`${URL}/`, data);
    return response.data;
}

export async function readoneBooking(booking_id: string): Promise<BookingResponse> {
    const response = await axios.get(`${URL}/${booking_id}`);
    return response.data;
}

export async function readallBookings(
    client_id?: string,
    technician_id?: string,
    status?: BookingStatus,
    booking_date?: Date,
    skip: number = 0,
    limit: number = 100
): Promise<BookingResponse[]> {
    const params: { [key: string]: string | number | BookingStatus | Date } = {};

    if (client_id !== undefined) {
        params.client_id = client_id;
    }
    if (technician_id !== undefined) {
        params.technician_id = technician_id;
    }
    if (status !== undefined) {
        params.status = status;
    }
    if (booking_date !== undefined) {
        params.booking_date = booking_date;
    }
    params.skip = skip;
    params.limit = limit;

    const response = await axios.get(URL, { params });
    return response.data;
}

export async function updateBooking(booking_id: string, data: BookingUpdate): Promise<BookingResponse> {
    const response = await axios.put(`${URL}/${booking_id}`, data);
    return response.data;
}

export async function deleteBooking(booking_id: string): Promise<boolean> {
    const response = await axios.delete(`${URL}/${booking_id}`);
    return response.data;
}
